File: 3회차/240_search_a_2d_matrix2/240_search_a_2d_matrix2_naive.py
'''
Array, Binary Search, DP

Binary search 적용 tip
어떤 정렬된 "배열"에서 "어떤 것"을 찾는다.
1. "어떤 것"을 정의해야 한다.
2. 조건에 따라 찾는 범위를 줄여 나간다. : "조건" 정의 필요 (기존: 중간보다 크다 혹은 작다)
3. 처음에는 양쪽 끝에서 시작, 그리고 그 어떤 것은 찾는 범위 내에서 가장 중간에서 시작.
4. 조건에 따라 찾는 범위를 반으로 줄여 나간다.

time complexity : O(logN)


'''

from typing import List

class Solution:
    def searchMatrix(self, matrix: List[List[int]], target: int) -> bool:
        '''
        모든 열 혹은 행에 대해 binary searching을 하는 것.
        '''
        
        m = len(matrix) # 행 
 
        for i in range(m):
            if self.binary_search(matrix[i], target):
                return True

        return False


    def binary_search(self, L : list, v) -> int:
        start, end = 0, len(L)-1

        while(start <= end):
            mid = (start + end) // 2

            if L[mid] == v:
                return True
            
            elif L[mid] < v:
                start = mid +1
            else:
                end = mid -1

        return False



def main():

    sol = Solution()

    matrix = [[1,4,7,11,15],[2,5,8,12,19],[3,6,9,16,22],[10,13,14,17,24],[18,21,23,26,30]]
    target = 20
    
    ans = sol.searchMatrix(matrix, target) 
    print(ans) #False


    m = [[-5]]
    t = -2 
    print(sol.searchMatrix(m, t)) # False

    m2 = [[-1,3]]
    t2= -1
    ans2 = sol.searchMatrix(m2, t2)
    print(ans2) # True


    m3 = [[5], [6]]
    t3 = 5
    ans3 = sol.searchMatrix(m3, t3)
    print(ans3)  # True



if __name__ == "__main__":
    main()

File: 3회차/240_search_a_2d_matrix2/240_search_a_2d_matrix2_naive.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>

// 이진 탐색 함수
bool binary_search(int* arr, int size, int target) {
    int start = 0;
    int end = size - 1;

    while (start <= end) {
        int mid = (start + end) / 2;

        if (arr[mid] == target) {
            return true;
        } else if (arr[mid] < target) {
            start = mid + 1;
        } else {
            end = mid - 1;
        }
    }
    return false;
}

// 행렬에서 target 찾기
bool searchMatrix(int** matrix, int matrixSize, int* matrixColSize, int target) {
    for (int i = 0; i < matrixSize; i++) {
        if (binary_search(matrix[i], matrixColSize[i], target)) {
            return true;
        }
    }
    return false;
}

int main() {
    // 행렬 동적 할당
    int rows = 5;
    int cols = 5;

    int** matrix = (int**)malloc(rows * sizeof(int*));
    for (int i = 0; i < rows; i++) {
        matrix[i] = (int*)malloc(cols * sizeof(int));
    }

    // 행렬 초기화
    int data[5][5] = {
        {1, 4, 7, 11, 15},
        {2, 5, 8, 12, 19},
        {3, 6, 9, 16, 22},
        {10, 13, 14, 17, 24},
        {18, 21, 23, 26, 30}
    };
    for (int i = 0; i < rows; i++) {
        for (int j = 0; j < cols; j++) {
            matrix[i][j] = data[i][j];
        }
    }

    // 각 행의 열 크기 저장
    int matrixColSize[5] = {5, 5, 5, 5, 5};

    // 타겟 값
    int target = 9;

    // 행렬에서 타겟 찾기
    if (searchMatrix(matrix, rows, matrixColSize, target)) {
        printf("Target %d found in matrix.\n", target);
    } else {
        printf("Target %d not found in matrix.\n", target);
    }

    // 메모리 해제
    for (int i = 0; i < rows; i++) {
        free(matrix[i]);
    }
    free(matrix);

    return 0;
}
